src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -140,6 +140,80 @@ object VariantSampleMatrix {
+    val colKeySchema = first.sSignature
+
+    datasets.indices.tail.foreach { i =>
+      val vds = datasets(i)
+      val ids = vds.sampleIds
+      val vas = vds.vaSignature
+      val gsig = vds.genotypeSignature
+      val vsig = vds.vSignature
+      val ssig = vds.sSignature
+
+      if (ssig != colKeySchema) {
+        fatal(
+          s"""cannot combine datasets with different column key schemata
+             |  Schema in datasets[0]: @1
+             |  Schema in datasets[$i]: @2""".stripMargin,
+          colKeySchema.toPrettyString(compact = true, printAttrs = true),
+          ssig.toPrettyString(compact = true, printAttrs = true)
+        )
+      } else if (vsig != rowKeySchema) {
+        fatal(
+          s"""cannot combine datasets with different row key schemata
+             |  Schema in datasets[0]: @1
+             |  Schema in datasets[$i]: @2""".stripMargin,
+          rowKeySchema.toPrettyString(compact = true, printAttrs = true),
+          vsig.toPrettyString(compact = true, printAttrs = true)
+        )
+      } else if (ids != sampleIds) {
+        fatal(
+          s"""cannot combine datasets with different column identifiers or ordering
+             |  IDs in datasets[0]: @1
+             |  IDs in datasets[$i]: @2""".stripMargin, sampleIds, ids)
+      } else if (wasSplit != vds.wasSplit) {
+        fatal(
+          s"""cannot combine split and unsplit datasets
+             |  Split status in datasets[0]: $wasSplit
+             |  Split status in datasets[$i]: ${ vds.wasSplit }""".stripMargin)
+      } else if (vas != vaSchema) {
+        fatal(
+          s"""cannot combine datasets with different row annotation schemata
+             |  Schema in datasets[0]: @1
+             |  Schema in datasets[$i]: @2""".stripMargin,
+          vaSchema.toPrettyString(compact = true, printAttrs = true),
+          vas.toPrettyString(compact = true, printAttrs = true)
+        )
+      } else if (gsig != genotypeSchema) {
+        fatal(
+          s"""cannot read datasets with different cell schemata
+             |  Schema in datasets[0]: @1
+             |  Schema in datasets[$i]: @2""".stripMargin,
+          genotypeSchema.toPrettyString(compact = true, printAttrs = true),
+          gsig.toPrettyString(compact = true, printAttrs = true)
+        )
+      }
+    }
+  }
+
+  def union(datasets: java.util.ArrayList[VariantSampleMatrix[Annotation, Annotation, Annotation]]): VariantSampleMatrix[Annotation, Annotation, Annotation] =
+    union(datasets.asScala.toArray)
+
+  def union(datasets: Array[VariantSampleMatrix[Annotation, Annotation, Annotation]]): VariantSampleMatrix[Annotation, Annotation, Annotation] = {
+    require(datasets.length >= 2)
+
+    checkDatasetSchemasCompatible(datasets)
+    val (first, others) = (datasets.head, datasets.tail)
+    first.copy(rdd = first.sparkContext.union(others.map(_.rdd)).toOrderedRDD(first.rdd.orderedPartitioner.kOk, first.rdd.vct))
+  }
```

Comment:
This loses the first dataset?

Reply:
Huh, yeah. How does this pass? My tests must be wrong.

Reply:
Yep, my test was bad and I should feel bad.